File: zenspider/2018/04.hs
import Control.Applicative (many)
import Data.Either         (fromRight)
import Data.List           (sort)
import Data.Map            (Map)
import Data.Ord            (comparing)
import Utils               (cleanStr, maximumVal, maximumValBy, minteract, occur)
import qualified Data.Map    as M
import qualified Text.Parsec as P

data Event = G Guard
           | S
           | W
           | E String
           deriving (Show, Eq, Ord)

data Time = T { _y :: Int, _m :: Int, _d :: Int, _h :: Int, _n :: Int }
  deriving (Show, Eq, Ord)

type Guard    = Int
type Minute   = Int
type Log      = (Time, Event)
type TimeCard = Map Minute Int -- min -> count
type Parser   = P.Parsec [Log] ()
type TimeCards = Map Guard TimeCard

parseLine :: String -> Log
parseLine s = (t, e)
  where (ts, rest) = splitAt 5 . words . cleanStr $ s
        [y,m,d,h,m_] = fmap read ts
        t            = T y m d h m_
        e            = case rest of
                         "falls":"asleep":_ -> S
                         "wakes":"up":_     -> W
                         "Guard":n:_        -> G $ read n
                         _                  -> error $ "bad: " ++ s

parseLogs :: [Log] -> TimeCards
parseLogs logs = fromRight M.empty $ P.parse go "" logs
  where
    go :: Parser TimeCards
    go = fmap occur . M.fromListWith (++) <$> many shift

    shift :: Parser (Guard, [Minute])
    shift = do
      (_, G g)   <- P.anyToken
      naps <- concat <$> many (P.try nap)
      return (g, naps)

    nap :: Parser [Minute]
    nap = do
      (T _ _ _ _ m0, S) <- P.anyToken
      (T _ _ _ _ m1, W) <- P.anyToken
      return [m0 .. m1 - 1]

solve1 :: TimeCards -> Int
solve1 tcs = guard * minute
  where
    (guard, tc) = maximumValBy (comparing sum) tcs
    (minute, _) = maximumVal tc

solve2 :: TimeCards -> Int
solve2 tcs = guard * minute
  where
    byMinute = fmap maximumVal tcs
    (guard, (minute, _)) = maximumValBy (comparing snd) byMinute

massage :: String -> TimeCards
massage = M.filter (not . null) . parseLogs . fmap parseLine . sort . lines

problem1 :: String -> String
problem1 = show . solve1 . massage

problem2 :: String -> String
problem2 = show . solve2 . massage

main :: IO ()
main = minteract [problem1, problem2]

File: zenspider/2018/04.rb
#!/usr/bin/env ruby -w

module Enumerable
  def count_by &blk
    group_by(&blk).transform_values(&:count)
  end
end

data    = Hash.new { |h,k| h[k] = [] }
date_re = '\d+-\d+-\d+ \d+:(\d+)'
current = nil

ARGF.read.lines.sort.each do |line|
  case line
  when /\[#{date_re}\] Guard #(\d+) begins shift/ then
    current = $2.to_i
  when /\[#{date_re}\] falls asleep/ then
    data[current] << $1.to_i
  when /\[#{date_re}\] wakes up/ then
    minutes = data[current].pop...$1.to_i
    data[current].concat minutes.to_a
    data[current].sort!                     # just for me
  end
end

# 04a:
guard, _count = data.transform_values { |v| v.count }.max_by(&:last)
data.transform_values! { |v| v.count_by(&:itself) } # { guard => { minute => count } }
minute, _count =  data[guard].max_by(&:last)

p guard * minute

# 04b:
# relying on previous transform_values!

# { guard => [max_minute, max_count] }
id, (min, _) = data.transform_values { |v| v.max_by(&:last) }.max_by { |_,(_,v)| v }
p id * min

__END__
[1518-11-01 00:00] Guard #10 begins shift
[1518-11-01 00:05] falls asleep
[1518-11-01 00:25] wakes up
[1518-11-01 00:30] falls asleep
[1518-11-01 00:55] wakes up
[1518-11-01 23:58] Guard #99 begins shift
[1518-11-02 00:40] falls asleep
[1518-11-02 00:50] wakes up
[1518-11-03 00:05] Guard #10 begins shift
[1518-11-03 00:24] falls asleep
[1518-11-03 00:29] wakes up
[1518-11-04 00:02] Guard #99 begins shift
[1518-11-04 00:36] falls asleep
[1518-11-04 00:46] wakes up
[1518-11-05 00:03] Guard #99 begins shift
[1518-11-05 00:45] falls asleep
[1518-11-05 00:55] wakes up
